File: src/GhostBot/controller/login_controller.py
from __future__ import annotations

import os
import threading
import time
from enum import Enum
from typing import TYPE_CHECKING

import pywintypes

from GhostBot import logger as _logger
from GhostBot.image_finder import ImageFinder
from GhostBot.lib.decorators import classproperty
from GhostBot.lib.talisman_ui_locations import UI_locations
from GhostBot.lib.utils import retry

if TYPE_CHECKING:
    from GhostBot.controller.bot_controller import BotClientWindow
    from GhostBot.controller.bot_controller import BotController
    from collections.abc import Callable
    from GhostBot.config import LoginDetailsConfigLoader


class LoginLock:
    _lock = threading.Lock()
    _locked_proc_id: str = ''
    _poll_frequency: float = 1
    logger = _logger.getChild('LoginLock')

    def __enter__(self):
        self.acquire(proc_id='context')

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.release()

    @classmethod
    def acquire(cls, proc_id: str, timeout: float = 30.0):
        start_time = time.time()
        while True:
            if cls._lock.acquire(timeout=1.0):
                cls._locked_proc_id = proc_id
                return cls
            
            if timeout is not None and (time.time() - start_time) > timeout:
                raise TimeoutError(f'LoginLock :: timed out waiting for lock to be unlocked after {timeout}s')
            cls.logger.debug('LoginLock :: waiting for lock to be unlocked, polling...')

    @classmethod
    def release(cls):
        cls._lock.release()
        cls._locked_proc_id = ''
        return cls

    @classproperty
    def locked(cls):
        return cls._lock.locked()

    @classproperty
    def unlocked(cls):
        return not cls._lock.locked()


class LoginController:
    _login_lock = LoginLock()

    def __init__(self, client: BotClientWindow, bot_controller: BotController):
        self.logger = _logger.getChild(self.__class__.__name__)
        self._bot_controller = bot_controller
        self._client = client
        self._config: LoginDetailsConfigLoader.CharDetails | None = None
        self._image_finder = ImageFinder(client)

    def set_config(self, config: LoginDetailsConfigLoader.CharDetails):
        self._config = config

    class LoginStage(Enum):
        enter_credentials = 0
        server_select = 1
        login_queue = 2
        character_select = 3
        success = 4

    @property
    def _enter_credentials(self):
        return bool(self._image_finder.find_ui_element(os.path.join(ImageFinder.misc_folder, 'login_main_page.bmp'),
                                                  threshold=0.6))

    @property
    def _server_select(self):
        return bool(self._image_finder.find_ui_element(os.path.join(ImageFinder.misc_folder, 'login_server_select.bmp')))

    @property
    def _login_queue(self):
        return bool(self._image_finder.find_ui_element(os.path.join(ImageFinder.misc_folder, 'login_queue.bmp')))

    @property
    def _character_select(self):
        if self._client.level:
            return False
        return not any((self._enter_credentials, self._server_select, self._login_queue))

    @property
    def current_stage(self):
        if self._enter_credentials:
            return self.LoginStage.enter_credentials
        elif self._server_select:
            return self.LoginStage.server_select
        elif self._login_queue:
            return self.LoginStage.login_queue
        elif self._character_select:
            return self.LoginStage.character_select
        else:
            return self.LoginStage.success

    def _handle_stage(self):
        match self.current_stage:
            case self.LoginStage.enter_credentials: self._handle_enter_credentials()
            case self.LoginStage.server_select: self._handle_server_select()
            case self.LoginStage.login_queue: self._handle_login_queue()
            case self.LoginStage.character_select: self._handle_character_select()
            case _: raise TypeError(f"unexpected stage {self.current_stage}")

    def _handle_enter_credentials(self):
        self.logger.debug("%s :: enter credentials", self._client.identifier)
        if self._config:
            self._client._name = self._config.char_name
            for _ in range(20):
                self._client.press_key('backspace')
            time.sleep(1)
            self._client.type_keys(self._config.username, char_only=True)
            self._client.press_key("tab")
            self._client.type_keys(self._config.password, char_only=True)
            self._client.press_key("enter")
            if not retry(lambda: self._server_select, 3, 2):
                self.logger.debug("%s :: login server is busy, restarting login process...", self._client.identifier)
                self._client.left_click((510, 335)) # 'login server is busy' dialog
                time.sleep(0.5)
                self._client.left_click((620, 390))  # username text box

    def _handle_server_select(self):
        self._login_lock.release()
        self.logger.debug("%s :: server select", self._client.identifier)
        if self._config and (server := self._config.server):
            self._client.left_click(UI_locations.server_select.get(server))
            time.sleep(1)
        self._client.left_click(UI_locations.server_select.ok)

    def _handle_login_queue(self):
        self.logger.debug("%s :: login queue, waiting 30s", self._client.identifier)
        time.sleep(25)

    def _handle_character_select(self):
        self.logger.debug("%s :: character select", self._client.identifier)
        def select_char(retry_count):
            self.logger.debug('%s :: waiting for game entered... (attempt %s)', self._client.identifier, retry_count)
            self._client.left_click(UI_locations.char_select_enter_game)
            time.sleep(1)
            self._client.initialize_pointers(force_reload=True)
            return self.current_stage == self.LoginStage.success

        if retry(select_char, 5, 1):
            self.logger.info("%s :: character logged in", self._client.identifier)
            self._client.post_login_setup()
        else:
            self.logger.info("%s :: character interrupted", self._client.identifier)
            self._client.left_click(UI_locations.char_select_interrupted_ok)

    def handle_login(self, callback: Callable):
        self.logger.debug("%s :: handle login", self._client.identifier)
        while self._client.level is None:
            if not self._bot_controller.running:
                self.logger.debug("%s :: bot controller not running, exiting...", self._client.identifier)
                return
            try:
                self._client.set_window_name()
                self._handle_stage()
                # Initialize the memory pointers, as they cant be set before login.
                self._client.set_window_name()
            except pywintypes.error as e:
                self.logger.exception(e)
                if e.args[2] == 'Invalid window handle.':
                    self.logger.error('%s :: ERROR with window handle.', self._client.identifier)
                    break  # jump out of this while loop, skipping the else block.
            except TypeError as e:
                self.logger.exception(e)
                break
            except Exception as e:
                self.logger.exception(e)
            time.sleep(1)
        else:
            self.logger.info("%s :: login handled", self._client.identifier)
            callback(True)
            return
        self.logger.error("%s :: login failed", self._client.identifier)
        callback(False)


if __name__ == "__main__":
    for i in LoginController.LoginStage:
        print(i.name)

File: kotlin-server/src/main/kotlin/com/ghostbot/controller/LoginController.kt
package com.ghostbot.controller

import com.ghostbot.ImageFinder
import com.ghostbot.config.LoginDetailsConfigLoader
import com.ghostbot.lib.UILocations
import com.ghostbot.lib.retry
import com.ghostbot.lib.retryWithCount
import com.ghostbot.lib.sleepSeconds
import com.ghostbot.rootLogger
import java.io.File
import java.util.logging.Level
import java.util.concurrent.locks.ReentrantLock

/**
 * Port of GhostBot/controller/login_controller.py.
 */

/**
 * Class-level lock so only one login procedure runs at a time.
 * Mirrors `LoginLock` (the Python `threading.Lock` classproperty behaviour).
 */
object LoginLock {
    private val lock = ReentrantLock(true)
    @Volatile
    private var lockedProcId: String = ""

    val logger = java.util.logging.Logger.getLogger("GhostBot.LoginLock")

    fun acquire(procId: String, timeout: Double = 30.0): LoginLock {
      val startTime = System.currentTimeMillis()
      while (true) {
          if (lock.tryLock(1L, java.util.concurrent.TimeUnit.SECONDS)) {
              lockedProcId = procId
              return this
          }
          if (timeout > 0 && (System.currentTimeMillis() - startTime) / 1000.0 > timeout) {
              throw java.util.concurrent.TimeoutException(
                  "LoginLock :: timed out waiting for lock to be unlocked after ${timeout}s"
              )
          }
          logger.log(Level.FINE, "LoginLock :: waiting for lock to be unlocked, polling...")
      }
    }

    fun release(): LoginLock {
        if (lock.isHeldByCurrentThread) {
            lock.unlock()
        }
        lockedProcId = ""
        return this
    }

    val locked: Boolean get() = lock.isLocked
    val unlocked: Boolean get() = !lock.isLocked

    /** Port of the `with LoginLock()` context manager. */
    inline fun <T> withLock(block: () -> T): T {
        acquire("context")
        try {
            return block()
        } finally {
            release()
        }
    }
}

class LoginController(
    private val client: BotClientWindow,
    private val botController: BotController,
) {
    companion object {
        val loginLock = LoginLock
    }

    enum class LoginStage {
        ENTER_CREDENTIALS,
        SERVER_SELECT,
        LOGIN_QUEUE,
        CHARACTER_SELECT,
        SUCCESS,
    }

    private val logger = java.util.logging.Logger.getLogger("GhostBot.LoginController")
    private var config: LoginDetailsConfigLoader.CharDetails? = null
    private val imageFinder = ImageFinder(client)
    private var serverBusyRetries = 0

    fun setConfig(config: LoginDetailsConfigLoader.CharDetails) {
        this.config = config
    }

    private val enterCredentials: Boolean
        get() = imageFinder.findUiElement(
            File(ImageFinder.miscFolder, "login_main_page.bmp").path, threshold = 0.6
        ) != null

    private val serverSelect: Boolean
        get() = imageFinder.findUiElement(
            File(ImageFinder.miscFolder, "login_server_select.bmp").path
        ) != null

    private val loginQueue: Boolean
        get() = imageFinder.findUiElement(
            File(ImageFinder.miscFolder, "login_queue.bmp").path
        ) != null

    private val characterSelect: Boolean
        get() {
            val level = client.level
            if (level != null) return false
            return !(enterCredentials || serverSelect || loginQueue)
        }

    val currentStage: LoginStage
        get() = when {
            enterCredentials -> LoginStage.ENTER_CREDENTIALS
            serverSelect -> LoginStage.SERVER_SELECT
            loginQueue -> LoginStage.LOGIN_QUEUE
            characterSelect -> LoginStage.CHARACTER_SELECT
            else -> LoginStage.SUCCESS
        }

    private fun handleStage() {
        when (currentStage) {
            LoginStage.ENTER_CREDENTIALS -> handleEnterCredentials()
            LoginStage.SERVER_SELECT -> handleServerSelect()
            LoginStage.LOGIN_QUEUE -> handleLoginQueue()
            LoginStage.CHARACTER_SELECT -> handleCharacterSelect()
            LoginStage.SUCCESS -> throw TypeError("unexpected stage SUCCESS")
        }
    }

    private class TypeError(msg: String) : RuntimeException(msg)

    private fun handleEnterCredentials() {
        logger.log(Level.FINE, "${client.identifier} :: enter credentials")
        val conf = config ?: return
        client.nameCache = conf.charName
        repeat(20) {
            client.pressKey("backspace")
        }
        sleepSeconds(1.0)
        client.typeKeys(conf.username, charOnly = true)
        client.pressKey("tab")
        client.typeKeys(conf.password, charOnly = true)
        client.pressKey("enter")

        if (!retry({ serverSelect }, retries = 3, delay = 2.0)) {
            serverBusyRetries++
            logger.log(
                Level.FINE,
                "${client.identifier} :: login server is busy, restarting login process... (retry $serverBusyRetries/3)"
            )
            client.leftClick(510 to 335) // 'login server is busy' dialog
            sleepSeconds(0.5)
            client.leftClick(620 to 390) // username text box

            if (serverBusyRetries >= 3) {
                logger.severe("${client.identifier} :: max 'server busy' retries reached, aborting login")
                throw RuntimeException("Max server busy retries reached")
            }
        } else {
            // Moved to server_select stage or succeeded, reset counter
            serverBusyRetries = 0
        }
    }

    private fun handleServerSelect() {
        loginLock.release()
        logger.log(Level.FINE, "${client.identifier} :: server select")
        val server = config?.server
        if (server != null && server.isNotEmpty()) {
            client.leftClick(UILocations.ServerSelect.byName[server] ?: UILocations.ServerSelect.ok)
            sleepSeconds(1.0)
        }
        client.leftClick(UILocations.ServerSelect.ok)
    }

    private fun handleLoginQueue() {
        logger.log(Level.FINE, "${client.identifier} :: login queue, waiting 30s")
        sleepSeconds(25.0)
    }

    private fun handleCharacterSelect() {
        logger.log(Level.FINE, "${client.identifier} :: character select")
        fun selectChar(retryCount: Int): Boolean {
            logger.log(Level.FINE, "${client.identifier} :: waiting for game entered... (attempt $retryCount)")
            client.leftClick(UILocations.charSelectEnterGame)
            sleepSeconds(1.0)
            client.initializePointers(forceReload = true)
            return currentStage == LoginStage.SUCCESS
        }

        if (retryWithCount(::selectChar, retries = 5, delay = 1.0)) {
            logger.info("${client.identifier} :: character logged in")
            client.postLoginSetup()
        } else {
            logger.info("${client.identifier} :: character interrupted")
            client.leftClick(UILocations.charSelectInterruptedOk)
        }
    }

    /**
     * Drives the login stages until the client's level becomes readable
     * (i.e. the character is in game), or the controller stops running.
     * Calls [callback] with true on success, false on failure.
     */
    fun handleLogin(callback: (Boolean) -> Unit) {
        logger.log(Level.FINE, "${client.identifier} :: handle login")
        var success = false
        while (client.level == null) {
            if (!botController.running) {
                logger.log(Level.FINE, "${client.identifier} :: bot controller not running, exiting...")
                return
            }
            try {
                client.setWindowName()
                handleStage()
                // Initialize the memory pointers, as they can't be set before login.
                client.setWindowName()
            } catch (e: Exception) {
                // pywintypes.error 'Invalid window handle.' / TypeError in the Python
                logger.log(Level.SEVERE, "login stage error for ${client.identifier}", e)
                break
            }
            sleepSeconds(1.0)
        }
        if (client.level != null) {
            logger.info("${client.identifier} :: login handled")
            success = true
        } else {
            logger.severe("${client.identifier} :: login failed")
        }
        callback(success)
    }
}
